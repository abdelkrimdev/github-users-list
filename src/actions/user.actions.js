import { userConstants } from '../constants/user.constants';
import { getAllUsers, getSingleUser } from '../services/github';

export const userActions = {
    getAll: (since = 0) => {
        return async (dispatch) => {
            try {
                dispatch(request());
                
                const payload = await getAllUsers(since);
                
                dispatch(success(payload));
            } catch (error) {
                dispatch(failure(error));
            }
        };

        function request() {
            return { type: userConstants.USER_ALL_REQUEST };
        }
        function success(users) {
            return { type: userConstants.USER_ALL_SUCCESS, users };
        }
        function failure(error) {
            return { type: userConstants.USER_ALL_FAILURE, error };
        }
    },
    getByUsername: (username) => {
        return async (dispatch) => {
            try {
                dispatch(request());
    
                const payload = await getSingleUser(username);
                
                dispatch(success(payload));
            } catch (error) {
                dispatch(failure(error));
            }
        };

        function request() {
            return { type: userConstants.USER_SPECIFIC_REQUEST };
        }
        function success(user) {
            return { type: userConstants.USER_SPECIFIC_SUCCESS, user };
        }
        function failure(error) {
            return { type: userConstants.USER_SPECIFIC_FAILURE, error };
        }
    }
};
