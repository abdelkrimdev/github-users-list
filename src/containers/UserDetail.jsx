import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Alert, Card, Col, Divider, Row, Spin } from 'antd';

import { userActions } from '../actions/user.actions';

export class UserDetail extends Component {
    static propTypes = {
        requesting: PropTypes.bool,
        user: PropTypes.object,
        error: PropTypes.object,
        getUser: PropTypes.func.isRequired
    }

    componentDidMount() {
        const { match: { params } } = this.props;
        this.props.getUser(params.username);
    }

    render() {
        const { requesting, user, error } = this.props;

        return (
            <div id="user-detail">
                { error && <Alert message="Error" description={ error.message } type="error" showIcon closable /> }
                <Row>
                    <Col span={8} offset={8}>
                        { requesting && <Spin /> }
                        { user &&
                            <Card cover={ <img alt={ user.login } src={ user.avatar_url } /> }>
                                <Card.Meta
                                    title={ <span><a href={ user.html_url }>{ user.login }</a><small>{ `#${user.id}` }</small></span> }
                                    description={ user.html_url }
                                />
                            </Card> }
                        <Divider />
                        <Link className="ant-btn" to="/">Show All</Link>
                    </Col>
                </Row>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    const { requesting, user, error } = state.specificUser;
    return {
        requesting,
        user,
        error
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        getUser: (username) => dispatch(userActions.getByUsername(username))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(UserDetail);
