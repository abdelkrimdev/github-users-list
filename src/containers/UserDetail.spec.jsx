import React from 'react';
import { createMockStore } from 'redux-test-utils';
import { shallow } from 'enzyme';

import UserDetail from './UserDetail';
import { userActions } from '../actions/user.actions';

describe('User Detail Component.', () => {
    let store;
    let wrapper;
    const username = 'someone';
    const match = { params: { username } };

    beforeEach(() => {
        store = createMockStore({
            specificUser: {}
        });
        wrapper = shallow(<UserDetail store={ store } match={ match } />).dive();
    });

    it('should be defined.', () => {
        expect(UserDetail).toBeDefined();
    });

    it('should render correctly.', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch get user action on user load the page.', () => {
        const expectedActions = [
            userActions.getByUsername(username)
        ];
        expect(store.isActionDispatched(userActions.getByUsername(username))).toBeTruthy();
    });

    it('should render a ​loading​ ​spinner on fetching the user data.', () => {
        store = createMockStore({
            specificUser: { requesting: true }
        });
        wrapper = shallow(<UserDetail store={ store } match={ match } />).dive();
        expect(wrapper.find('Spin')).toHaveLength(1);
    });

    it('should render a ​Card displaying the user data.', () => {
        store = createMockStore({
            specificUser: {
                user: { "id": 0, "login": username, "avatar_url": "", "html_url": "" }
            }
        });
        wrapper = shallow(<UserDetail store={ store } match={ match } />).dive();
        expect(wrapper.find('Card')).toHaveLength(1);
    });

    it('should render a Link to​ ​go​ ​back​ ​to​ ​the​ ​initial​ ​list.', () => {
        expect(wrapper.find('Link')).toHaveLength(1);
    });
});
