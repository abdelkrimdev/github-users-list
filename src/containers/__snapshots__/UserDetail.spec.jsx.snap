// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Detail Component. should render correctly. 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserDetail
    error={undefined}
    getUser={[Function]}
    match={
        Object {
            "params": Object {
              "username": "someone",
            },
          }
    }
    requesting={undefined}
    store={
        Object {
            "dispatch": [Function],
            "getAction": [Function],
            "getActions": [Function],
            "getState": [Function],
            "isActionDispatched": [Function],
            "isActionTypeDispatched": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dispatch": [Function],
              "getAction": [Function],
              "getActions": [Function],
              "getState": [Function],
              "isActionDispatched": [Function],
              "isActionTypeDispatched": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": undefined,
          }
    }
    user={undefined}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        <Row
          gutter={0}
>
          <Col
                    offset={8}
                    span={8}
          >
                    <Divider />
                    <Link
                              className="ant-btn"
                              replace={false}
                              to="/"
                    >
                              Show All
                    </Link>
          </Col>
</Row>,
      ],
      "id": "user-detail",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Col
            offset={8}
            span={8}
>
            <Divider />
            <Link
                        className="ant-btn"
                        replace={false}
                        to="/"
            >
                        Show All
            </Link>
</Col>,
          "gutter": 0,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              undefined,
              undefined,
              <Divider />,
              <Link
                className="ant-btn"
                replace={false}
                to="/"
>
                Show All
</Link>,
            ],
            "offset": 8,
            "span": 8,
          },
          "ref": null,
          "rendered": Array [
            undefined,
            undefined,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Show All",
                "className": "ant-btn",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Show All",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <Row
            gutter={0}
>
            <Col
                        offset={8}
                        span={8}
            >
                        <Divider />
                        <Link
                                    className="ant-btn"
                                    replace={false}
                                    to="/"
                        >
                                    Show All
                        </Link>
            </Col>
</Row>,
        ],
        "id": "user-detail",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Col
              offset={8}
              span={8}
>
              <Divider />
              <Link
                            className="ant-btn"
                            replace={false}
                            to="/"
              >
                            Show All
              </Link>
</Col>,
            "gutter": 0,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                undefined,
                undefined,
                <Divider />,
                <Link
                  className="ant-btn"
                  replace={false}
                  to="/"
>
                  Show All
</Link>,
              ],
              "offset": 8,
              "span": 8,
            },
            "ref": null,
            "rendered": Array [
              undefined,
              undefined,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Show All",
                  "className": "ant-btn",
                  "replace": false,
                  "to": "/",
                },
                "ref": null,
                "rendered": "Show All",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
